### PR TITLE
[WIP] Snapshot record video

### PR DESCRIPTION
**Pull request by JulesMoorhouse**
### Checklist
- [x] I've run `bundle exec rspec` from the root directory to see all new and existing tests pass
- [x] I've followed the _fastlane_ code style and run `bundle exec rubocop -a` to ensure the code style is valid
- [x] I've read the [Contribution Guidelines](https://github.com/fastlane/fastlane/blob/master/CONTRIBUTING.md)
- [x] I've updated the documentation if necessary.

### Motivation and Context
This change add Snapshot to the ability to record videos of iOS applications, using the same method used to take screenshots.

### Description
Added ability to record of video within the simulator

### CODE REVIEW

**Comment by Jules2010** [comment]
@laullon please will you sign the CLA again here please. I'm going to try and get this pushed through without the tcp listener.

**Comment by laullon** [comment]
sure, I’ll sign it later.
What’s your plan?

**Comment by Jules2010** [comment]
Cheers.

I’ve not done any ruby yet.

Initially just get something working, I have 6 languages and 8 devices to produce videos for, I need something to automate this task.

However I may need to add some extra features, not sure what au be needed yet.

**Comment by Jules2010** [comment]
I timed my existing snapshot lane last night and found I'm going to need to remove some the functionality to make it fit within 30 seconds.

I'm hoping to do all this in-app, without having to pause the video.

I've already got a start and finish slide, with fading, that going to be around 7 seconds.

I figure I can dismiss view controllers without animation, then push without animation, to speed up the transition between several screens.

At least that's my plan.

Otherwise I may need to add some kind of pause then fade to fastlane.

**Comment by Jules2010** [comment]
I'm struggling, can't figure out how to call the start / stop methods :( https://github.com/fastlane/fastlane/pull/11744/commits/c6458f0ca66b79c76738d1ed4fee38247fbc4e86

**Comment by laullon** [comment]
call the start / stop methods from where? UITest?

**Comment by laullon** [comment]
I signed it!

**Comment by Jules2010** [comment]
Thanks.

Yes from the UITest any help appreciated.

**Comment by laullon** [comment]
sure... That was the purpose of the "tcp listener"/"web server" ;)

**Comment by Jules2010** [comment]
Yeah. I’m just keen to get it merged, then hopefully it will get more features and discussion.

Any ideas how I need to add the start / stop in to the swift files, I couldn’t see it on snapshot or simulator. 

Maybe it needs exposing somehow? 

I’m a bit out of my depth with ruby.

**Comment by Jules2010** [comment]
I signed it!

**Review by KrauseFx — CHANGES_REQUESTED** [review]
Hey, cool to see some movement in this again. Please make sure to avoid adding a web server to this, as it adds a lot of complexity to the code base. I'm using `Request Change` to make sure this doesn't get merged until it's ready, it's still WIP, right?

**Comment by Jules2010** [comment]
@KrauseFx I’ve taken the web server stuff out. 

However I’m not familiar with ruby and am struggling to add calls to the start and stop functions to the swift can you help with this small change please?

**Comment by laullon** [comment]
@KrauseFx any recommendation on how to communicated the UITest with the Fastlane process to start/stop the recording ?

I did used the web server because it was very easy of implement, I also tested to use a flag file, but that was not ok because I did used a 3rd party library to watch for file system changes.

**Comment by getaaron** [comment]
There are a few other options:

- Swift can write to `STDOUT` using `print()`. Ruby can read from this process, potentially using `pty` or `IO`. Some [helpful-looking Stack Overflow discussion here](https://stackoverflow.com/q/1154846/1445366).
- The Swift code can write to a file and we can use something like [filewatcher](https://github.com/thomasfl/filewatcher) to watch for changes within fastlane.
- Since we know we're running on macOS, there are many other [macOS-specific interprocess communication methods](https://developer.apple.com/library/content/referencelibrary/GettingStarted/GS_InterapplicationCommunication/_index.html) available (XPC, copy and paste, notification center, etc.)

I think the `STDOUT` approach (using `print()` and `IO`) is least likely to require additional dependencies, but it may not be the most elegant.

@KrauseFx any thoughts here?

File: snapshot/lib/assets/SnapfileTemplate [hunk 1]
```diff
@@ -23,8 +23,15 @@
 # Where should the resulting screenshots be stored?
 # output_directory "./screenshots"
 
-# remove the '#' to clear all previously generated screenshots before creating new ones
-# clear_previous_screenshots true
+
+# Where should the resulting videos be stored?
```
Comment (janpio) [inline snapshot/lib/assets/SnapfileTemplate]: this comment should take into account that by default no video will be create. So probably somethign like

File: snapshot/lib/assets/SnapfileTemplate [hunk 2]
```diff
@@ -23,8 +23,15 @@
...
+# video_output_directory "./videos"
+
+# clear_previous_screenshots true # remove the '#' to clear all previously generated screenshots before creating new ones
+
+# Choose which project/workspace to use
```
Comment (janpio) [inline snapshot/lib/assets/SnapfileTemplate]: unrelated new feature or just new in SnapfileTemplate?

File: snapshot/lib/snapshot.rb [hunk 3]
```diff
@@ -17,3 +17,46 @@
 require_relative 'snapshot/simulator_launchers/simulator_launcher'
 require_relative 'snapshot/simulator_launchers/simulator_launcher_xcode_8'
 require_relative 'snapshot/module'
+
+require 'fastlane_core'
+
+require 'open3'
+
+module Snapshot
```
Comment (janpio) [inline snapshot/lib/snapshot.rb]: a `Snapshot` module should be placed in `snapshot/module.rb` relative from here,

Comment (janpio) [inline snapshot/lib/snapshot.rb]: When you rebase onto current `master`, this module will alredy exist in `snapshot/module` relative from here.

Comment (janpio) [reply snapshot/lib/snapshot.rb]: If not, and this is just a useless copy, one would have to make sure if there are no important changes here.

Comment (janpio) [reply snapshot/lib/snapshot.rb]: Quick comparison shows that they are indeed identical. This is probably just an artifact (that has to be fixed before a merge).

File: snapshot/lib/snapshot/options.rb [hunk 4]
```diff
@@ -7,6 +7,7 @@ module Snapshot
   class Options
     def self.available_options
       output_directory = (File.directory?("fastlane") ? "fastlane/screenshots" : "screenshots")
+      video_output_directory = (File.directory?("fastlane") ? "fastlane/videos" : "videos")
```
Comment (janpio) [inline snapshot/lib/snapshot/options.rb]: does this folder already exist by default?

File: snapshot/lib/snapshot/simulator_launchers/simulator_launcher.rb [hunk 5]
```diff
@@ -134,6 +137,39 @@ def execute(retries = 0, command: nil, language: nil, locale: nil, launch_args:
                                               end)
     end
 
+    def start_recording(dir_name, device, name)
```
Comment (janpio) [inline snapshot/lib/snapshot/simulator_launchers/simulator_launcher.rb]: maybe add `video` to the method name to make it clearer what this is?

File: snapshot/lib/snapshot/simulator_launchers/simulator_launcher.rb [hunk 6]
```diff
@@ -134,6 +137,39 @@ def execute(retries = 0, command: nil, language: nil, locale: nil, launch_args:
...
+      device_udid = TestCommandGenerator.device_udid(device)
+      pid = @recording_pid[device_udid]
+      return unless pid.nil?
```
Comment (janpio) [inline snapshot/lib/snapshot/simulator_launchers/simulator_launcher.rb]: what exactly does this do and why?

File: snapshot/lib/snapshot/simulator_launchers/simulator_launcher.rb [hunk 7]
```diff
@@ -134,6 +137,39 @@ def execute(retries = 0, command: nil, language: nil, locale: nil, launch_args:
                                               end)
     end
 
+    def start_recording(dir_name, device, name)
+      device_udid = TestCommandGenerator.device_udid(device)
```
Comment (janpio) [inline snapshot/lib/snapshot/simulator_launchers/simulator_launcher.rb]: extract into well named method as it is repeated below

File: snapshot/lib/snapshot/simulator_launchers/simulator_launcher.rb [hunk 8]
```diff
@@ -134,6 +137,39 @@ def execute(retries = 0, command: nil, language: nil, locale: nil, launch_args:
...
+                                                         UI.error("Caught error... #{return_code}")
+                                                         UI.error("Caught output... #{output}")
+                                                       end)
+      end
+    end
+
+    def stop_recording(device)
+      device_udid = TestCommandGenerator.device_udid(device)
+      pid = @recording_pid[device_udid]
+      return if pid.nil?
```
Comment (janpio) [inline snapshot/lib/snapshot/simulator_launchers/simulator_launcher.rb]: what exactly does this do and why?

File: snapshot/lib/snapshot/simulator_launchers/simulator_launcher.rb [hunk 9]
```diff
@@ -134,6 +137,39 @@ def execute(retries = 0, command: nil, language: nil, locale: nil, launch_args:
...
+                                                       end,
+                                                error: proc do |output, return_code|
+                                                         ErrorHandler.handle_test_error(output, return_code)
+                                                         UI.error("Caught error... #{return_code}")
+                                                         UI.error("Caught output... #{output}")
+                                                       end)
+      end
+    end
+
+    def stop_recording(device)
```
Comment (janpio) [inline snapshot/lib/snapshot/simulator_launchers/simulator_launcher.rb]: also add `video` to method name maybe?

File: snapshot/lib/snapshot/simulator_launchers/simulator_launcher.rb [hunk 10]
```diff
@@ -66,6 +67,8 @@ def take_screenshots_simultaneously
             device_batches = launcher_config.devices.map { |d| [d] }
           end
 
+          @recording_pid = {}
```
Comment (janpio) [inline snapshot/lib/snapshot/simulator_launchers/simulator_launcher.rb]: should this also be in the `_xcode_8` version of this file?

File: snapshot/lib/snapshot/simulator_launchers/simulator_launcher_xcode_8.rb [hunk 11]
```diff
@@ -101,6 +102,35 @@ def execute(command: nil, language: nil, locale: nil, device_type: nil, launch_a
                                                   UI.crash!("Too many errors... no more retries...")
                                                 end
                                               end)
+      stop_recording
+    end
+
+    def start_recording(dir_name, device_type, name)
```
Comment (janpio) [inline snapshot/lib/snapshot/simulator_launchers/simulator_launcher_xcode_8.rb]: if this is identical to above, can this be moved outside of these files somehow and unified?

File: snapshot/lib/snapshot/simulator_launchers/simulator_launcher_xcode_8.rb [hunk 12]
```diff
@@ -89,6 +89,7 @@ def execute(command: nil, language: nil, locale: nil, device_type: nil, launch_a
                                             loading: "Loading...",
                                               error: proc do |output, return_code|
                                                 ErrorHandler.handle_test_error(output, return_code)
+                                                stop_recording
```
Comment (janpio) [inline snapshot/lib/snapshot/simulator_launchers/simulator_launcher_xcode_8.rb]: should this be in the other, non `_xcode_8` version of this file?

**Comment by Jules2010** [comment]
Hi @janpio 

I'd just like to mention that this pull request as it stands doesn't provide any additional functionality as it stands. It's based on PR https://github.com/fastlane/fastlane/pull/10121 I'm not 100% sure about your questions.

However this pull request has the http listener code removed and therefore doesn't have any means of triggering the video recording.

I started this PR at that point, in the hope that someone would step in and advise and contribute to it's completion. I understand that @laullon has moved on from his PR with other projects and doesn't have time to contribute further.

For information purposes, I have created a new branch https://github.com/Jules2010/fastlane/tree/snapshot-record-video-with-audio which includes @laullon  initial work and QuickTime Player to also record sound.

I hope this can move forward.

Jules.

**Comment by janpio** [comment]
Yes, I got that, saw the other PR as well. 

I tried to understand what is already here and see if this is in an optimal state to solve the remaining problem - because I really like the possible functionality here! Using snapshot to record videos of the app would really be an awesome feature.

Maybe @laullon can also take a look at my comments?

**Comment by getaaron** [comment]
@Jules2010 How is this work coming along? Were you able to get it working for your project locally without the HTTP server?

**Comment by Jules2010** [comment]
@getaaron unfortunately not, I didn’t write the code originally. But I was able to add sound via QuickTime and use ffmpeg to join the audio and video together.

However my code isn’t pretty and has hardcoded paths.

I did mention to @laullon on twitter but he obviously didn’t find the time to help further.

**Comment by Lausbert** [comment]
If someone is looking for inspiration for a pull request without a http server: [https://github.com/Lausbert/Snaptake](https://github.com/Lausbert/Snaptake) :)

**Comment by Jules2010** [comment]
@Lausbert does this add sound to the videos too? My hack uses QuickTime and merged them together.

**Comment by getaaron** [comment]
@lausbert I really like this approach; what do you think about setting this up as a fastlane plugin? We could add a comment about it in the Snapfile and link to it in the docs.

**Comment by Lausbert** [comment]
@Jules2010 I haven't tested it so far, since all I needed to do in my specific case was to add a jingle later on with ffmpeg.

@getaaron Sounds great, but what about the changes in snapshotHelper.swift? Did I miss anything or is it impossible to integrate them by adding a plugin?

**Comment by Jules2010** [comment]
@Lausbert the simulator currently doesn’t provide sound. I filed a radar and was told it was a feature request.

**Comment by janpio** [comment]
Closing this PR as I don't expect a solution to suddenly pop up.

But I created a new issue https://github.com/fastlane/fastlane/issues/14362 which links to this PR and also @Lausbert 's working solution - maybe we can restart a discussion there and find a solution that can be elegantly added to Snapshot.

Thanks all that were involved here!

**Pull request closed without merging**